Fix ACC_MODULE class verification

theresa-m opened the pull request:
* Class file is only a module if ACC_MODULE is the only flag set.
* Remove ACC_MODULE from CFR_CLASS_ACCESS_MASK. A valid class cannot have this flag set.

Fixes: #740 

Signed-off-by: Theresa Mammarella <Theresa.T.Mammarella@ibm.com>

Peer reviewer: @babsingh

babsingh commented on runtime/bcutil/cfreader.c:
```diff
@@ -1174,7 +1174,7 @@ readPool(J9CfrClassFile* classfile, U_8* data, U_8* dataEnd, U_8* segment, U_8*
 			}
 
-			if (!J9_ARE_ALL_BITS_SET(classfile->accessFlags, CFR_ACC_MODULE)) {
+			if (classfile->accessFlags != CFR_ACC_MODULE) {
```
generally, we do bit operations/checks on`accessFlags`. better to add this comment over here as well:
```
	/* Java 9 Spec: If the ACC_MODULE flag is set in the access_flags item, then no
	 * other flag in the access_flags item may be set. If other flags are set, then this
	 * file is not a module.
```

babsingh commented on runtime/bcutil/cfreader.c:
```diff
@@ -2201,8 +2201,10 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
+	/* check annotations and interfaces for invalid flags. If the ACC_INTERFACE flag is set,
+	 * the ACC_FINAL, ACC_SUPER, and ACC_ENUM flags set must not be set.
+	 */
+	if ((value & CFR_ACC_INTERFACE) && (value & CFR_INTERFACE_FORBIDDEN_ACCESS_MASK))
```
better to use the macro: `J9_ARE_ALL_BITS_SET(value, CFR_ACC_INTERFACE | CFR_INTERFACE_FORBIDDEN_ACCESS_MASK)`

babsingh commented on runtime/bcutil/cfreader.c:
```diff
@@ -2442,8 +2445,13 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 	}
 
+	/* Java 9 Spec: If the ACC_MODULE flag is set in the access_flags item, then no
+	 * other flag in the access_flags item may be set. If other flags are set, then this
```
`then no other flag in the access_flags item may be set`: may be -> should be?

theresa-m commented:
Thanks @babsingh. I applied your changes.

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2442,8 +2449,13 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 	if ((vmVersionShifted >= BCT_Java9MajorVersionShifted)
-			&& (J9_ARE_ALL_BITS_SET(classfile->accessFlags, CFR_ACC_MODULE))) {
+			&& (classfile->majorVersion >= 53)
+			&& (classfile->accessFlags == CFR_ACC_MODULE)) {
```
I don't understand this change. It seems it should be checking if the ACC_MODULE bit is set as in the original code, but not if only the ACC_MODULE bit is set.

theresa-m replied:
In the Java 9 spec a file is considered a module only if ACC_MODULE is the only access flag set. I added this to double check that modules are not loaded like classes and to match Oracles behaviour for this scenario.

pshipton replied:
Understood, but it seems you need to check if the module bit is set in the class, not if its the only bit set. Or is there there a check somewhere else that checks if the module bit is set? The code doesn't match the description in the comment.

theresa-m replied:
Ah the case where ACC_MODULE is set but its not an invalid module?

I also found the rule `If the ACC_INTERFACE flag is not set, ACC_MODULE may not be set.` in the new spec. I will add a check for this in that should address that case.

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2201,8 +2205,10 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
+	/* check annotations and interfaces for invalid flags. If the ACC_INTERFACE flag is set,
+	 * the ACC_FINAL, ACC_SUPER, and ACC_ENUM flags set must not be set.
+	 */
+	if (J9_ARE_ALL_BITS_SET(value, CFR_ACC_INTERFACE) && J9_ARE_ANY_BITS_SET(value, CFR_INTERFACE_FORBIDDEN_ACCESS_MASK))
```
ACC_MODULE is also invalid for Java 9

theresa-m replied:
Thanks. Looks like the Java 9 spec I've been using is a bit out of date.

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2442,8 +2449,13 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
+	 */
 	if ((vmVersionShifted >= BCT_Java9MajorVersionShifted)
-			&& (J9_ARE_ALL_BITS_SET(classfile->accessFlags, CFR_ACC_MODULE))) {
+			&& (classfile->majorVersion >= 53)
```
Is this check necessary after checking vmVersionShifted?

theresa-m replied:
These are not both needed.

theresa-m commented:
Updated checks to java 9 spec:
* If ACC_INTERFACE is set, ACC_MODULE must not be set
* If the ACC_INTERFACE flag is not set, ACC_MODULE may not be set

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2195,18 +2195,35 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
 
+	/* Java 9 spec: If the ACC_INTERFACE flag is not set, ACC_MODULE may not be set. */
+	if ((flags & BCT_MajorClassFileVersionMask) >= BCT_Java9MajorVersionShifted) {
+		if ((value & CFR_ACC_MODULE) && (! (value & CFR_ACC_INTERFACE))) {
```
Since this is new code, please use the J9 BITS macros.

pshipton replied:
Doesn't this conflict with the below check that if CFR_ACC_INTERFACE is set then CFR_ACC_MODULE cannot be set?

theresa-m replied:
Yes - I'll these checks are correct. I've now simplified them into one check.

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2195,18 +2195,35 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
-		goto _errorFound;
+		 if ((((flags & BCT_MajorClassFileVersionMask) >= BCT_Java9MajorVersionShifted)
+				 && J9_ARE_ANY_BITS_SET(value, CFR_INTERFACE_FORBIDDEN_ACCESS_MASK_9))
+			|| J9_ARE_ANY_BITS_SET(value, CFR_INTERFACE_FORBIDDEN_ACCESS_MASK)) {
```
Please move the `) {` to the next line for this multi-line if statement. It should line up with the if.

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2442,8 +2460,7 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 
-	if ((vmVersionShifted >= BCT_Java9MajorVersionShifted)
-			&& (J9_ARE_ALL_BITS_SET(classfile->accessFlags, CFR_ACC_MODULE))) {
+	if ((classfile->majorVersion >= 53) && (classfile->accessFlags == CFR_ACC_MODULE)) {
```
Why the change to use classfile->majorVersion instead of vmVersionShifted? If its required, please add comments to explain it, and create or use an existing defined constant for 53. 

I still don't think its correct to check `accessFlags == CFR_ACC_MODULE` rather than checking if the CFR_ACC_MODULE bit is set.

theresa-m replied:
Tests were failing in some cases using vmVersionShifted because a class file with version below 53 can be loaded by a Java 9 vm and this check should not apply.

pshipton commented:
There used to be a check that .class'es with the module attribute must only have the ACC_MODULE bit set, which I thought was valid, but has since been removed.

theresa-m commented:
I think I've just been making this worse by adding more checks - I checked that Hotspot is throwing the same error message for modules for all cases of a Java 9 class file with ACC_MODULE set, so in my latest push I simply moved the check further up in verification so it checks sooner, and removed ACC_MODULE from CFR_CLASS_ACCESS_MASK that I added in a fix earlier. Since ACC_MODULE is not part of a class that is valid I realized this doesn't make a lot of sense.

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2378,6 +2375,7 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 	/* protect the callers with default values */
 	if ((flags & BCT_MajorClassFileVersionMask) == 0) {
```
I don't think this check is correct. It appears the MajorClassFileVersion portion of the flags is initialized further down in this function. It can be moved higher, but the classfile->majorVersion needs to be read in first. We shouldn't be initializing a default of Java 9.

theresa-m replied:
Just pushed a change to address this.

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2130,11 +2130,16 @@ checkAttributes(J9CfrClassFile* classfile, J9CfrAttribute** attributes, U_32 att
+
+	/* protect the callers with default values */
+	if (0 == max_allowed_version) {
+		max_allowed_version = BCT_Java9MajorVersionShifted >> BCT_MajorClassFileVersionMaskShift;
```
This needs to be fixed not to default to anything. We need to consistently know which version of Java the VM is currently running, since it supports multiple versions.

theresa-m replied:
Removed this check and fixed tests / cfdumper.

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2171,11 +2176,7 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
-	} else {
-		value = classfile->accessFlags & CFR_CLASS_ACCESS_MASK;
-	}
+	value = classfile->accessFlags & CFR_CLASS_ACCESS_MASK;
```
The vmVersionShifted parameter to the function is no longer needed.

pshipton replied:
The vmVersionShifted parameter to the checkPool() function is no longer needed.

pshipton commented on runtime/bcutil/cfreader.c:
```diff
@@ -2397,6 +2388,13 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 	NEXT_U16(classfile->minorVersion, index);
 	NEXT_U16(classfile->majorVersion, index);
+
+	/* Make sure that following verification uses the class file version number. VM version is
```
Since this code was moved, it should change to say "the structure and static verification below" instead of "following verification"

theresa-m replied:
Just updated the comment.

pshipton commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -2017,6 +2017,7 @@ typedef struct J9BCTranslationData {
 #define BCT_InitInstanceFields  19
 #define BCT_ErrConstantPoolMapTooSmall  14
 #define BCT_InitCopyDoubles  5
+#define BCT_JavaMaxMajorVersionShifted BCT_Java10MajorVersionShifted
```
Please flip the order so its 8, 9, 10, Max. Leave a blank line after Max to make it more noticable, and add a comment before 8 to say something like when adding a new major version update BCT_JavaMaxMajorVersionShifted to the maximum allowed value.
